handler/file/retrieve.go

```diff
@@ -45,11 +71,150 @@ func (DefaultHandler) RetrieveFileHandler(
 		return nil, file.FileName(), time.Unix(0, file.LastModifiedNano), errors.WithStack(err)
 	}
 
-	// TODO: if we cannot open the file from the file system, use Filecoin retrieval
 	seeker, obj, err := storagesystem.Open(rclone, ctx, file.Path)
 	if err != nil {
-		return nil, file.FileName(), time.Unix(0, file.LastModifiedNano), errors.WithStack(err)
+		// we have no local copy, so we have to return a filecoin based reader
+		//nolint:nilerr
+		return &filecoinReader{
+			ctx:       ctx,
+			db:        db,
+			retriever: filecoinRetriever,
+			size:      file.Size,
+			id:        id,
+		}, file.FileName(), time.Unix(0, file.LastModifiedNano), nil
 	}
 
 	return seeker, file.FileName(), obj.ModTime(ctx), nil
 }
+
+// io.ReadSeekCloser implementation that reads from remote singularity
+type filecoinReader struct {
+	ctx       context.Context
+	db        *gorm.DB
+	retriever FilecoinRetriever
+	offset    int64
+	size      int64
+	id        uint64
+}
+
+func (r *filecoinReader) Read(p []byte) (int, error) {
+	logger.Infof("buffer size: %v", len(p))
+
+	buf := bytes.NewBuffer(p)
+	buf.Reset()
+
+	if r.offset >= r.size {
+		return 0, io.EOF
+	}
+
+	// Figure out how many bytes to read
+	readLen := int64(len(p))
+	remainingBytes := r.size - r.offset
+	if readLen > remainingBytes {
+		readLen = remainingBytes
+	}
+
+	fileRanges, err := findFileRanges(r.db, r.id, r.offset, r.offset+readLen)
```

Comment:
This has some room to optimize to reduce db and network requests. Even during sequential read, the length of the buffer `p` increases gradually so you may end up calling `findFileRanges` and `findProviders` multiple times.

How about finding all file ranges during the first time when read() is called so we can reduce the database load.

Reply:
And probably providers too.